src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -229,47 +231,61 @@ await foreach (ConfigurationSetting setting in _client.GetConfigurationSettingsA
                         LabelFilter = loadOption.LabelFilter
                     };
 
-                    // Load all key-values with the null label.
                     await CallWithRequestTracing(async () =>
                     {
                         await foreach (ConfigurationSetting setting in _client.GetConfigurationSettingsAsync(selector, CancellationToken.None).ConfigureAwait(false))
                         {
-                            data[setting.Key] = setting;
+                            serverData[setting.Key] = setting;
                         }
                     }).ConfigureAwait(false);
                 }
 
                 // Block current thread for the initial load of key-values registered for refresh that are not already loaded
-                await Task.Run(() => LoadKeyValuesRegisteredForRefresh(data).ConfigureAwait(false).GetAwaiter().GetResult()).ConfigureAwait(false);
+                await Task.Run(() => LoadKeyValuesRegisteredForRefresh(serverData).ConfigureAwait(false).GetAwaiter().GetResult()).ConfigureAwait(false);
+
+                if (_options.OfflineCache != null)
+                {
+                    _options.OfflineCache.Export(_options, JsonSerializer.Serialize(serverData));
+                }

```

Comment:
Could we keep the existing behavior to export only after the configuration load is complete? This will ensure that we don't delay loading of settings just to save the offline copy.

Reply:
Since we call `SetData` only once after the catch block, we don't know if this data was loaded from AppConfig or from OfflineCache. Moving the export here ensures that if there is an exception, and if we end up loading data from existing offline cache - we don't need to export same data back to offline cache.

Reply:
The existing code handles this case without exporting the offline cache data again in case of an exception. I wouldn't change that behavior since it affects the performance observed by the customer. One approach could be to only export if `cachedData` is `null`.

Reply:
Done